sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -570,6 +571,33 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 
+object SimplifyConditionalInPredicate extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f @ Filter(cond, _) => f.copy(condition = simplifyConditional(cond))
+  }
+
+  private def simplifyConditional(e: Expression): Expression = e match {
+    case cw @ CaseWhen(branches, elseValue) if cw.dataType == BooleanType && branches.size == 1 &&
+        elseValue.forall(_.semanticEquals(FalseLiteral)) =>
+      val (whenVal, thenVal) = branches.head
+      And(whenVal, thenVal)
+    case i @ If(pred, trueVal, FalseLiteral) if i.dataType == BooleanType =>
+      And(pred, trueVal)
+    case e if e.dataType == BooleanType =>
+      e
+    case e =>
+      val message = "Expected a Boolean type expression in simplifyConditional, " +
+        s"but got the type `${e.dataType.catalogString}` in `${e.sql}`."
+      if (Utils.isTesting) {
```

Comment:
In this case, shall we raise exception always?

Reply:
I plan merge this logic to `ReplaceNullWithFalseInPredicate` in future commit: 
https://github.com/apache/spark/blob/09edff51ad5bce7b6f2b4b982d6df550a37d502e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala#L103-L111